src/gui/layertree/qgslayertreemapcanvasbridge.cpp

```diff
@@ -160,7 +160,7 @@ void QgsLayerTreeMapCanvasBridge::setCanvasLayers( QgsLayerTreeNode *node, QList
       if ( QgsGroupLayer *groupLayer = group->groupLayer() )
       {
         canvasLayers << groupLayer;
-        continue;
+        break;
```

Comment:
I don't fully understand this fix -- if the child node is a group layer, then the continue here will already ensure that the call to setCanvasLayers on line 166 isn't called. But by using 'break' here then we risk missing nodes (eg a node which contains two child nodes, where the first is a group and the other is just a normal child layer)

Reply:
It's not the child node that is a group ; it's `node` itself. But it could probably be rewritten to just skip the children loop when `node` is a `QgsLayerTreeGroup` (that is linked to a QgsGroupLayer)

Reply:
ooooo -- I think the actual bug is that it should be:

```if ( QgsLayerTreeGroup *group = QgsLayerTree::toGroup( child ) )```

Reply:
Yes, I tried that, it crashed, and I got lazy.  :sweat_smile:

So the problem lies with :

` if ( QgsLayerTreeGroup *group = QgsLayerTree::toGroup( node ) ) `

This does not work to check the validity since `QgsLayerTree::toGroup` performs a static_cast.
This didn't crash before, since this code was reached only if `node` actually had children.

I added a `QgsLayerTree::isGroup` check.

Why not use `dynamic_cast` for `QgsLayerTree::toGroup` and `QgsLayerTree::toLayer` though?